File: ts/tests/conformance-helpers.ts
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import type {
  Body,
  GrpcInteraction,
  HeaderMap,
  HttpInteraction,
  WsInteraction,
} from "../src/types.js";

export const CONFORMANCE_ROOT = join(dirname(fileURLToPath(import.meta.url)), "..", "..");

function sortKeys<T extends object>(value: T): T {
  return Object.fromEntries(
    Object.entries(value).sort(([left], [right]) => (left < right ? -1 : 1)),
  ) as T;
}

export function canonicalBody(value: Body): object {
  return value.type === "none"
    ? { type: "none" }
    : { type: value.type, content: value.content };
}

function headers(value: HeaderMap): HeaderMap {
  return sortKeys(value);
}

export function canonicalCassette(cassette: {
  version: number;
  interactions: HttpInteraction[];
  grpcInteractions: GrpcInteraction[];
  wsInteractions: WsInteraction[];
}): object {
  return {
    version: cassette.version,
    http: cassette.interactions.map((interaction) => ({
      method: interaction.request.method,
      uri: interaction.request.uri,
      requestHeaders: headers(interaction.request.headers),
      requestBody: canonicalBody(interaction.request.body),
      status: interaction.response.status,
      responseHeaders: headers(interaction.response.headers),
      responseBody: canonicalBody(interaction.response.body),
      recordedAt: interaction.recordedAt,
    })),
    grpc: cassette.grpcInteractions.map((interaction) => ({
      method: interaction.request.method,
      metadata: headers(interaction.request.metadata),
      requestBody: canonicalBody(interaction.request.body),
      statusCode: interaction.response.statusCode,
      statusMessage: interaction.response.statusMessage,
      responseMetadata: headers(interaction.response.metadata),
      responseBody: canonicalBody(interaction.response.body),
      jsonDebug: interaction.jsonDebug ?? null,
      recordedAt: interaction.recordedAt,
    })),
    ws: cassette.wsInteractions.map((interaction) => ({
      uri: interaction.uri,
      headers: headers(interaction.headers),
      frames: interaction.frames.map((frame) => ({
        direction: frame.direction,
        frameType: frame.frameType,
        body: canonicalBody(frame.body),
        offsetMs: frame.offsetMs,
      })),
      recordedAt: interaction.recordedAt,
    })),
  };
}

File: tests/conformance_helpers.py
from __future__ import annotations

from typing import Any

from cassetter._core import Body, Cassette


def canonical_body(body: Body) -> dict[str, Any]:
    if body.body_type == "binary":
        return {"type": "binary", "content": body.content.hex()}
    if body.body_type == "none":
        return {"type": "none"}
    return {"type": body.body_type, "content": body.content}


def canonical_headers(headers: dict[str, list[str]]) -> dict[str, list[str]]:
    return {key: list(value) for key, value in sorted(headers.items())}


def canonical_cassette(cassette: Cassette) -> dict[str, Any]:
    return {
        "version": cassette.version,
        "http": [
            {
                "method": interaction.request.method,
                "uri": interaction.request.uri,
                "requestHeaders": canonical_headers(interaction.request.headers),
                "requestBody": canonical_body(interaction.request.body),
                "status": interaction.response.status,
                "responseHeaders": canonical_headers(interaction.response.headers),
                "responseBody": canonical_body(interaction.response.body),
                "recordedAt": interaction.recorded_at,
            }
            for interaction in cassette.interactions
        ],
        "grpc": [
            {
                "method": interaction.request.method,
                "metadata": canonical_headers(interaction.request.metadata),
                "requestBody": canonical_body(interaction.request.body),
                "statusCode": interaction.response.status_code,
                "statusMessage": interaction.response.status_message,
                "responseMetadata": canonical_headers(interaction.response.metadata),
                "responseBody": canonical_body(interaction.response.body),
                "jsonDebug": interaction.json_debug,
                "recordedAt": interaction.recorded_at,
            }
            for interaction in cassette.grpc_interactions
        ],
        "ws": [
            {
                "uri": interaction.uri,
                "headers": canonical_headers(interaction.headers),
                "frames": [
                    {
                        "direction": frame.direction,
                        "frameType": frame.frame_type,
                        "body": canonical_body(frame.body),
                        "offsetMs": frame.offset_ms,
                    }
                    for frame in interaction.frames
                ],
                "recordedAt": interaction.recorded_at,
            }
            for interaction in cassette.ws_interactions
        ],
    }
